crates/bevy_reflect/src/serde/mod.rs

```diff
@@ -181,4 +185,197 @@ mod tests {
 
         assert!(expected.reflect_partial_eq(&result).unwrap());
     }
+
+    mod type_data {
+        use super::*;
+        use crate::from_reflect::FromReflect;
+        use crate::serde::serialize_reflect::{ReflectSerializeReflect, SerializeReflect};
+        use crate::serde::{DeserializeReflect, ReflectDeserializeReflect};
+        use crate::{ReflectFromReflect, TypePath};
+        use bevy_reflect_derive::reflect_trait;
+        use core::fmt::{Debug, Formatter};
+        use serde::de::{SeqAccess, Visitor};
+        use serde::ser::SerializeSeq;
+        use serde::{Deserializer, Serializer};
+        use std::sync::Arc;
+
+        #[reflect_trait]
+        trait Enemy: Reflect + Debug {
+            fn hp(&self) -> u8;
+        }
+
+        // This is needed to support Arc<dyn Enemy>
+        impl TypePath for dyn Enemy {
+            fn type_path() -> &'static str {
+                "dyn bevy_reflect::serde::tests::type_data::Enemy"
+            }
+
+            fn short_type_path() -> &'static str {
+                "dyn Enemy"
+            }
+        }
+
+        #[derive(Reflect, Debug)]
+        #[reflect(Enemy)]
+        struct Skeleton(u8);
+
+        impl Enemy for Skeleton {
+            fn hp(&self) -> u8 {
+                self.0
+            }
+        }
+
+        #[derive(Reflect, Debug)]
+        #[reflect(Enemy)]
+        struct Zombie(u8);
+
+        impl Enemy for Zombie {
+            fn hp(&self) -> u8 {
+                self.0
+            }
+        }
+
+        #[derive(Reflect, Debug)]
+        struct Level {
+            name: String,
+            enemies: EnemyList,
+        }
+
+        #[derive(Reflect, Debug)]
+        #[reflect(SerializeReflect, DeserializeReflect)]
+        struct EnemyList(Vec<Arc<dyn Enemy>>);
```

Comment:
nit: Should this be clearer with `Box<dyn Enemy>`? Why the `Arc` here?

Reply:
It would be clearer but unfortunately `Box<dyn Reflect>` isn't reflectable yet

Reply:
Oh, and `Arc<dyn Reflect>` is? I didn't know that. 🤯

Reply:
It's over `Arc<T: Send + Sync>` and only as a `reflect_value` type: https://github.com/bevyengine/bevy/blob/5619bd09d1668dc3b0f14833bb1c3d7cffa9f9ef/crates/bevy_reflect/src/impls/std.rs#L207

It was added to enable some reflection stuff for Assets v2. We might revisit it once we know what to do with smart pointers in general.

Reply:
Thanks for the details. Reflection of pointers is a hard topic. 😔